### PR TITLE
Allow disabling browse mode entirely or stay in focus mode when explicitly enabled

### LINKED ISSUE #8716: Feature request:  Lock focus-mode:
Overview: It is sometimes preferred to default to interacting with the Web in focus mode.  For example, partially-sighted people using a magnifier with NVDA may prefer to start in, and stay in Focus Mode. 
Therefore, the locked-focus setting would maintain focus-mode even when opening new browser windows, restarting the browser or possibly even when switching browsers.

 First, please create a key command to lock focus mode, default could be NVDA+ctrl+space.

Second:  Please create a configuration file setting to lock focus mode, The user could place this in either a global or a triggered-configuration.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #8716 

### Summary of the issue:
People who use NVDA combined with a magnifier sometimes prefer not to use browse mode on the web. There are also cases where a sighted assistant operates the computer without browse mode, in which case is it frustrating that browse mode is activated for every new page load.

### Description of how this pull request fixes the issue:
1. Added an option to control on load browse mode activation. 

    * For web pages: the current behavior; activate browse mode on the web, and in Excel/Word where browse mode is also supported, do not enable it by default. This is mostly based on obj.shouldCreateTreeInterceptor
    * For all supported documents: even enables browse mode for word and excel documents, i.e. everywhere where browse mode is supported. This is generally not what a user should do, but might make sense in scope of configuration profiles.
    * Never: Only enable browse mode when explicitly activated using NVDA+space. This is what someone would use who relies primarily on browse mode.

2. Added an option to remember manually activated focus mode on page load. This option does the following:

	* For browse mode on the web, do not enable browse mode again when pass through has been explicitly disabled by the user with NVDA+space. Thus:
		+ When searching on Google and the edit box is activated with space to enable focus mode, pressing enter will load a new page with browse mode enabled, since this wasn't an explicit disable.
		+ When searching on Google and the edit box is activated with NVDA+space to enable focus mode (i.e. explicit pass through), pressing enter will load a new page with browse mode disabled. The user will have to re-enable browse mode again with NVDA+space.
	* Currently, this function respects the case where auto browse mode is set to never. Thus, if browse mode has to be explicitly enabled with NVDA+space after a page load, loading a new page will always result in browse mode being disabled by default. Basically, this means that this option does nothing when auto browse mode is set to never.
	* This option is completely ignored for cases where browse mode has to be invoked manually. This is because, when you have always on browse mode and you enable pass through in Word, you don't want to have Firefox load a new page with browse mode disabled.

### Testing performed:
Tested the several scenarios as outlined above.

### Known issues with pull request:
* The UX might have to be changed in some ways. This is why I didn't yet change the user guide.

### Change log entry:
* New features
    + IN NVDA's browse mode settings, you can now disable browse mode completely, only making it available manually by pressing NVDA+space. (#8716)

### CODE REVIEW

File: source/treeInterceptorHandler.py [hunk 1]
```diff
@@ -13,26 +13,62 @@
...
 	#If this object already has a treeInterceptor, just return that and don't bother trying to create one
 	ti=obj.treeInterceptor
 	if not ti:
-		if not obj.shouldCreateTreeInterceptor:
-			return None
 		try:
 			newClass=obj.treeInterceptorClass
 		except NotImplementedError:
 			return None
+		# Import late to avoid circular input.
```
Comment (derekriemer) [inline source/treeInterceptorHandler.py]: circular input > circular import

File: source/treeInterceptorHandler.py [hunk 2]
```diff
@@ -13,26 +13,62 @@
...
+				(obj.shouldCreateTreeInterceptor and config.conf['virtualBuffers']['createTIOnPageLoad'] == TI_CREATE_NEVER) or
+				(not obj.shouldCreateTreeInterceptor and config.conf['virtualBuffers']['createTIOnPageLoad'] != TI_CREATE_ALWAYS) or 
+				# When report pass through on load is on and the tree interceptor class is a browse mode one,
+				# we should not create a tree interceptor if pass through has earlier been enabled explicitly.
+				(config.conf['virtualBuffers']['respectPassThroughOnPageLoad'] and
+					issubclass(newClass, BrowseModeTreeInterceptor) and
+					lastDisableAutoPassThrough
+				)
+			):
+				return None
```
Comment (derekriemer) [inline source/treeInterceptorHandler.py]: Possible improvement: return None is unnecessary. Return works.

Comment (derekriemer) [reply source/treeInterceptorHandler.py]: Update: the style here is return None. You should probably comply with the surrounding code as you are doing.

**Review by derekriemer — DISMISSED** [review]

**Comment by michaelDCurran** [comment]
I sort of understand the use case for disabling browseMode treeInterceptors all together. Though I would certainly only refuse to create browseModeTreeInterceptors, not just any treeInterceptors. As TreeInterceptors could be used for different purposes, such as focus filtering.
I don't believe we should have the "always" option at all. I think it over complicates this code, and I  don't follow the use case.
The staying in focus mode feature makes sense to me in regards to a new page loading, but where I worry is alt+tabbing to another browser window, or switching tabs. this will be a very visible UX change. If this part is not needed for the original problem with magnifiers, I'd split it into two separate PRs.

**Comment by leonardder** [comment]
> I sort of understand the use case for disabling browseMode treeInterceptors all together. Though I would certainly only refuse to create browseModeTreeInterceptors, not just any treeInterceptors. As TreeInterceptors could be used for different purposes, such as focus filtering.

This makes sense and I will address this.

> I don't believe we should have the "always" option at all. I think it over complicates this code, and I don't follow the use case.

To be honest, I only added this possibility to have a clear opposite of the never option, and it was trivial to implement. I could imagine that there are cases where someone wants to force browse mode in a certain application (i.e. in a text review configuration profile for Word), but it is indeed somewhat complicating things. Should I just remove this ability altogether?

> The staying in focus mode feature makes sense to me in regards to a new page loading, but where I worry is alt+tabbing to another browser window, or switching tabs. this will be a very visible UX change. 

Note that the staying in focus mode only applies to the last object with a tree interceptor that has focus. To clarify:

1. When on google.com and going to nvaccess.org with browse mode disabled, nvaccess.org will load without browse mode.
2. There are two tabs, one with babbage.com (browse mode enabled) and one with google.com (browse mode disabled). When on google.com and going to nvaccess.org with browse mode disabled, nvaccess.org will load without browse mode. However, when switching back to babbage.com which had browse mode enabled, browse mode will be enabled again.

> If this part is not needed for the original problem with magnifiers, I'd split it into two separate PRs.

Nope, it isn't.

Having discussed this, it might make sense to close this pr and open two separate ones, where the force disabling browse mode one takes priority for me.

**Comment by michaelDCurran** [comment]
Yep. My vote would be two separate PRs, and remove the "always" option 
all together.

**Pull request closed without merging**